### ISSUE TITLE
Ingest generated spreadsheet processes do not link to multiple dissociation protocols

### ISSUE THREAD

**Issue by Wkt8** [ebi-ait/dcp-ingest-central#942]
**Describe the bug**
Dissociation and enrichment protocols (and possibly others) are not linked appropriately in the Ingest-generated-xls spreadsheet for a project

**To Reproduce**
Steps to reproduce the behaviour:
1. Go to 'https://contribute.data.humancellatlas.org/submissions/detail?uuid=87da3563-99af-4e55-9c30-2732b0af07d4&project=77c13c40-a598-4036-807f-be09209ec2dd'
2. Click on 'Download xls spreadsheet'
3. In the spreadsheet, go to the 'cell_suspension' tab and find process with uuid `59e90e1e-0098-43e3-8297-c6563d287f1d`
4. The process should be linked to only one dissociation protocol.
5. In Ingest UI, scroll down to 'Processes'
6. Find process with uuid `59e90e1e-0098-43e3-8297-c6563d287f1d` - it should be linked to multiple dissociation protocols

**Expected behaviour**
The generated excel spreadsheet should have processes which are linked to multiple dissociation protocols, as shown in Ingest UI 

**Screenshots / Logs**
> If applicable, add screenshots or log files to help explain your problem.

**Links**
> If applicable, link to relevant docs or Slack conversations.

**Environment**
- [x] prod
- [ ] staging
- [ ] dev

**Browser**
- [x] Chrome
- [ ] Safari
- [ ] IE
- [ ] Other (please specify)

**Comment by ESapenaVentura** [comment]
Found the error, the problem is in this line:

https://github.com/ebi-ait/ingest-client/blob/master/hca_ingest/downloader/flattener.py#L156

tl;dr: When multiple protocols of the same concrete type (e.g. dissociation_protocols) are present in linkings, alongside other types (e.g. enrichment protocols), the function tries to group them based on that concrete type. However, the `groupBy` function only works on **ordered lists**. In the example above, the protocol list for all processes going into cell suspensions has the following order:
1. Dissociation protocol
1. Enrichment protocol
1. Dissociation protocol
Which ends up deleting the first dissociation protocol when iterating.

I have written a bit of a snippet on a solution, but not sure if it's the best one:

```python
def sort_by_concrete_type(prot):
    return prot.schema.concrete_type
for concrete_type, inputs_iter in groupby(sorted(entity_of_interest.protocols, key=sort_by_concrete_type), lambda entity: entity.schema.concrete_type):
```

**Comment by anu-shiva** [comment]
@ESapenaVentura to open a PR for review by @amnonkhen 
Possible investigation if this happened with other datasets.

**Comment by ESapenaVentura** [comment]
PR to fix this: https://github.com/ebi-ait/ingest-client/pull/69

**Comment by ESapenaVentura** [comment]
Corrected according to feedback; added testing and a slight adjustment on iterator management

**Comment by gabsie** [comment]
hey @ESapenaVentura and @amnonkhen  - is this now resolved, should it be closed?

**Comment by ESapenaVentura** [comment]
Hi @gabsie - No, the PR has not yet been merged. I will merge on Tuesday and check how to push the new version to the broker and exporter

**Comment by ESapenaVentura** [comment]
Waiting on access to pipy - Will deploy version 2.8.1, then merge PR, then create PRs in the [Ingest graph validator](https://github.com/ebi-ait/ingest-graph-validator), [Ingest broker](https://github.com/ebi-ait/ingest-broker) and [Ingest exporter](https://github.com/ebi-ait/ingest-exporter) repositories to update the dependency
